### DISCUSSION TITLE
Here to answer questions and consider requests!

### DISCUSSION THREAD

**Discussion by pmcculler** [pmcculler/sd-dynamic-javascript#2] [Announcements]
Just drop a note with your compliment, complaint, or... neutral statement?

**Comment by pmcculler** [comment]
We are up to 1,039 unique cloners, going at about 100 new unique clones a day slowly dropping, which I expect is quite normal. Anyway, I hope it's in actual use and that you are getting what you hoped for from it.

**Comment by curiousjp** [comment]
Thanks for putting this together - I read it as part of writing my own script to do something similar with Python. Very much appreciated.

Incidentally, although I never actually got a chance to run your solution (I didn't have a webdriver handy), can I ask you a question about it? 

Does the replacement of p.all_prompts in your `process()` function with a single element list based on the transformed first item from p.all_prompts (i.e. all_prompts = [prompt_out]) cause this script to fail when batch size is otherwise than 1? My read of `setup_prompts` in processing.py is that the all_prompts list has to be equal in length to batch_size * n_iter.  Thanks again!!

**Reply by pmcculler** [reply]
I wondered about that too. Figured it would break right away; but it hasn't failed on me yet. Not at all sure what's up with that. LMK if you figure it out, please.

**Reply by curiousjp** [reply]
So having tested it a bit here, if I have e.g. a batch size of three, truncating the p.all_prompts list to a single element causes it to silently stop after processing the first image in the batch - no error, it just stops. So my code iterates over the full list and then packs the processed versions back into 'p', i.e. the relevant part of my `process()` function looks like this: 
```Python
prompt_count = len(p.all_prompts)
for i in range(prompt_count):
    pos_prompt = p.all_prompts[i]
    neg_prompt = p.all_negative_prompts[i]
    try:
        global_context = {}
        local_context = { "batch_index": i }
        if "%%" in pos_prompt:
            pos_prompt, global_context, local_context = _rewrite_prompt(pos_prompt, global_context, local_context)
        if "%%" in neg_prompt:
            neg_prompt, global_context, local_context = _rewrite_prompt(neg_prompt, global_context, local_context)
    except Exception as e:
        logging.exception(e)
        pos_prompt = str(e)
        neg_prompt = str(e)
    p.all_prompts[i] = pos_prompt
    p.all_negative_prompts[i] = neg_prompt
```
Works well for me anyway as I would sometimes like my code to behave differently on each image depending on e.g. the position in the generation list or the value of pseudorandom choices.

**Reply by pmcculler** [reply]
So you iterate over all the p.all_prompts & negative prompts and modify each of them individually. Funny, I was modifying a fork of webui-cutoff yesterday and this was what I wound up doing. Figured I would see about porting the idea back into the JS extension and here you are telling me I should do just that. Well, thanks! Good to know indeed.
Is your extension public, or published? What's it do?

**Reply by curiousjp** [reply]
It’s very similar to this one, actually (I even ended up using your %% semantics for marking blocks) but it executes a blob of Python instead of JavaScript.

Python’s exec takes (and returns) dictionaries representing before and after local and global variables, so state is propagated between the positive and negative prompts, and I can also preload the state with some generation information. However, because it’s not so convenient to return() in the context of exec(), the value of a designated return value is read back from the local state dictionary to specify the block’s results. 

It’s not public at the moment because I do no sanitisation of the code and it’s quite unsafe to inject Python code directly into a Python program (fixed local namespace or not), but I’d be happy to gist you a copy of it later if you’d like.

**Reply by pmcculler** [reply]
I would very much like to see & try that out! Funny, I also made an embedded C# extension the other day, but it suffers from the same level of security problem, and on top of that you need a tool installed that I think is only found in the Visual Studio suite. I think I can work around that by writing my own little text-to-Rosalyn bridge, but haven't gotten around to it... maybe because I have yet to see someone complain they can't use C# to script their prompts, lolz. It looks from the snippet above like you were stuffing local variables in for scripts to access, which is pretty neat and something I might well adopt in the near future. Thanks for the inspiration.
I may need to make the %% delimiter something that can be changed in settings. 🤔
So one thing I had to make decision about was interacting with other extensions, some of which make be executed before or after. What did you decide? Like, do you want your embedded python extension to run before Dynamic Prompts, or after?

**Reply by curiousjp** [reply]
You know, I use so few prompt rewriting extensions that I hadn't given it much thought (particularly as at the moment I'm now building the prompts more or less outside of A111 and then bringing them in with the 'batch from file' feature.) I'm actually not even sure how the extensions handler schedules it... Anyway, I'm glad if it was a useful conversation in terms of ideas - I've tagged you on that private gist - hopefully you'll get a ping off it and see that very little has changed from your own code!

**Reply by pmcculler** [reply]
Got it and thank you! I talked with someone who was automating bringing in prompts via the built-in API the other day. I may just start doing that myself, possibly make a quick extension in a code editor for it, so, external prompt generation something like you're doing.
Oddly enough, extensions are executed in alphabetical order. So, because I wanted to be able to execute Javascript both before and after Dynamic Prompts... so I could do this {\_\_my_javascript_code\_\_}... I wound up creating a second, similar extension with a different name, alphabetically after DP's. It runs any code that's appeared from Dynamic Prompts with the standard delimiter, and it has its own delimiter as well, as I had a scenario where I wanted to run the code _only_ after Dynamic Prompts.
